RFC: Authorization support (#133)
modelcontextprotocol/specification: docs/specification/draft/basic/authorization.md

### 1.3 Standards Compliance

This authorization mechanism is based on established specifications listed below, but
implements a selected subset of their features to ensure security and interoperability
while maintaining simplicity:

- [OAuth 2.1 IETF DRAFT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12)
- OAuth 2.0 Authorization Server Metadata
  ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
- OAuth 2.0 Dynamic Client Registration Protocol
  ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591))

## 2. Authorization Flow

### 2.1 Overview

1. MCP auth implementations **MUST** implement OAuth 2.1 with appropriate security
   measures for both confidential and public clients.

2. MCP auth implementations **SHOULD** support the OAuth 2.0 Dynamic Client Registration
   Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. MCP servers **SHOULD** and MCP clients **MUST** implement OAuth 2.0 Authorization
   Server Metadata ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414)). Servers
   that do not support Authorization Server Metadata **MUST** follow the default URI
   schema.

### 2.2 Basic OAuth 2.1 Authorization

> When authorization is required and not yet prooven by the client, servers **MUST**

Reviewer: ```suggestion
When authorization is required and not yet proven by the client, servers **MUST**
```

respond with _HTTP 401 Unauthorized_.

Clients initiate the
[OAuth 2.1 IETF DRAFT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12)
authorization flow after receiving the _HTTP 401 Unauthorized_.

The following demonstrates the basic OAuth 2.1 for public clients using PKCE.

```mermaid
sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server

    C->>M: MCP Request
    M->>C: HTTP 401 Unauthorized
    Note over C: Generate code_verifier and code_challenge
    C->>B: Open browser with authorization URL + code_challenge
    B->>M: GET /authorize
    Note over M: User logs in and authorizes
    M->>B: Redirect to callback URL with auth code
    B->>C: Callback with authorization code
    C->>M: Token Request with code + code_verifier
    M->>C: Access Token (+ Refresh Token)
    C->>M: MCP Request with Access Token
    Note over C,M: Begin standard MCP message exchange
```

### 2.3 Server Metadata Discovery

For server capability discovery:

- MCP clients _MUST_ follow the OAuth 2.0 Authorization Server Metadata protocol defined
  in [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).

> - MCP server _SHOULD_ support follow the OAuth 2.0 Authorization Server Metadata

Reviewer: ```suggestion
- MCP server _SHOULD_ follow the OAuth 2.0 Authorization Server Metadata
```

  protocol.
- MCP servers that do not support the OAuth 2.0 Authorization Server Metadata protocol,

>   _MUST_ support fallback URLs..

Reviewer: ```suggestion
  _MUST_ support fallback URLs.
```

The discovery flow is illustrated below:

```mermaid
sequenceDiagram
    participant C as Client
    participant S as Server

    C->>S: GET /.well-known/oauth-authorization-server
    alt Discovery Success
        S->>C: 200 OK + Metadata Document
        Note over C: Use endpoints from metadata
    else Discovery Failed
        S->>C: 404 Not Found
        Note over C: Fall back to default endpoints
    end
    Note over C: Continue with authorization flow
```

#### 2.3.1 Server Metadata Discover Headers

MCP clients _SHOULD_ include the header `MCP-Protocol-Version: <protocol-version>` during

> Server Metadata Discover to allow MCP server to respond based on MCP protocol version.

Reviewer: ```suggestion
Server Metadata Discovery to allow the MCP server to respond based on the MCP protocol version.
```

**MUST** use the following default endpoint paths relative to the server's base URL:

| Endpoint               | Default Path | Description                          |
| ---------------------- | ------------ | ------------------------------------ |
| Authorization Endpoint | /authorize   | Used for authorization requests      |
| Token Endpoint         | /token       | Used for token exchange & refresh    |
| Registration Endpoint  | /register    | Used for dynamic client registration |

Clients **MUST** first attempt to discover endpoints via the metadata document before
falling back to default paths. When using default paths, all other protocol requirements
remain unchanged.

### 2.3 Dynamic Client Registration

MCP authorization makes use of the
[OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591)
to allow MCP clients to obtain OAuth client IDs without user interaction.

- MCP clients **SHOULD** support
  [OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591)
- MCP servers with non-localhost redirect URIs **SHOULD** support Dynamic Client
  Registration.

> - MCP servers with localhost redirect URIs **OPTIONALLY** support Dynamic Client

Reviewer: ```suggestion
- MCP servers with localhost redirect URIs **MAY** support Dynamic Client
```

(nit for grammar—alternatively, "may optionally")

  Registration.

> Note that are not required to support Dynamic Client Registration. MCP clients that do

Reviewer: ```suggestion
Note that it is not required to support Dynamic Client Registration. MCP clients that do
```

not support Dynamic Client Registration need to provide alternative ways to obtain a
client id (and if applicable client secret).

#### 2.3.1 Localhost Redirect URIs

When using localhost redirect URIs (http://localhost:{port} or http://127.0.0.1:{port}),
clients:

- Dynamic registration is **OPTIONAL** (a client ID is not required)
- **MAY** proceed directly to authorization

> - **MUST NOT** require client secrets

Reviewer: I'm still confused by this bullet. What does it mean that "clients must not require client secrets?"

registration. This provides a standardized way for clients to automatically register with
new servers, which is crucial for MCP because:

- Clients cannot know all possible servers in advance
- Manual registration would create friction for users
- It enables seamless connection to new servers
- Servers can implement their own registration policies

### 2.4 Authorization Flow Steps

The complete Authorization flow proceeds as follows:

```mermaid
sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server

    C->>M: GET /.well-known/oauth-authorization-server
    alt Server Supports Discovery
        M->>C: Authorization Server Metadata
    else No Discovery
        M->>C: 404 (Use default endpoints)
    end

    alt Non-Localhost Redirect URI
        C->>M: POST /register
        M->>C: Client Credentials
    end

    Note over C: Generate PKCE Parameters
    C->>B: Open browser with authorization URL + code_challenge
    B->>M: Authorization Request
    Note over M: User /authorizes
    M->>B: Redirect to callback with authorization code
    B->>C: Authorization code callback
    C->>M: Token Request + code_verifier

>     M->>C: Access Token + Refresh Token

Reviewer: ```suggestion
    M->>C: Access Token (+ Refresh Token)
```

    B->>T: Authorization Request
    Note over T: User authorizes
    T->>B: Redirect to MCP Server callback
    B->>M: Authorization code
    M->>T: Exchange code for token
    T->>M: Third-party access token
    Note over M: Generate bound MCP token
    M->>B: Redirect to MCP Client callback
    B->>C: MCP authorization code
    C->>M: Exchange code for token
    M->>C: MCP access token
```

#### 2.9.3 Session Binding Requirements

MCP servers implementing third-party authorization **MUST**:

1. Maintain secure mapping between third-party tokens and issued MCP tokens
2. Validate third-party token status before honoring MCP tokens
3. Implement appropriate token lifecycle management
4. Handle third-party token expiration and renewal

#### 2.9.4 Security Considerations

When implementing third-party authorization, servers **MUST**:

1. Validate all redirect URIs
2. Securely store third-party credentials
3. Implement appropriate session timeout handling
4. Consider security implications of token chaining
5. Implement proper error handling for third-party auth failures

## 3. Best Practices

#### 3.1 Local clients as Public OAuth 2.1 Clients

> We strongly recommend that local clients implement OAuth 2.1 as public client:

Reviewer: ```suggestion
We strongly recommend that local clients implement OAuth 2.1 as a public client:
```

1. Utilizing code challenges (PKCE) for authorization requests to prevent interception
   attacks
2. Implementing secure token storage appropriate for the local system
3. Following token refresh best practices to maintain sessions
4. Properly handling token expiration and renewal

#### 3.2 Authorization Metadata Discovery

> We strongly recommend, that all clients implement metadata discovery. This reduces the

Reviewer: ```suggestion
We strongly recommend that all clients implement metadata discovery. This reduces the
```

> need for users to provide endpoints either manually or fallback to the defined defaults.

Reviewer: ```suggestion
need for users to provide endpoints manually or for fallback to the defined defaults.
```
